intra/android/tun2socks.go

```diff
@@ -79,3 +85,23 @@ func NewDoHTransport(url string, ips string, protector protect.Protector, auth d
 	dialer := protect.MakeDialer(protector)
 	return doh.NewTransport(url, split, dialer, auth, listener)
 }
+
+func copyUntilEOF(dst, src io.ReadWriteCloser) {
+	log.Printf("[debug] start relaying traffic [%s] -> [%s]", src, dst)
+	defer log.Printf("[debug] stop relaying traffic [%s] -> [%s]", src, dst)
+
+	const commonMTU = 1500
+	buf := make([]byte, commonMTU)
+	defer dst.Close()
+	for {
+		_, err := io.CopyBuffer(dst, src, buf)
+		if err == nil || isErrClosed(err) {
+
+			return
+		}
+	}
+}
+
+func isErrClosed(err error) bool {
+	return errors.Is(err, os.ErrClosed) || errors.Is(err, fs.ErrClosed) || errors.Is(err, network.ErrClosed)
```

Comment:
Do we return `network.ErrClosed` on Reads? I thought we would just return `ErrClosed` on writes, and return `EOF` on reads.

Reply:
It's not us, it's the tun device (os.File) that doesn't conform with go's rules, it will return `os.ErrClosed`. But I expanded to more similar `ErrClosed` errors here.